includes/class-wc-payment-gateway-wcpay.php

```diff
@@ -1365,6 +1372,28 @@ public function update_customer_with_order_data( $order, $customer_id, $is_test_
 		$this->customer_service->update_customer_for_user( $customer_id, $user, $customer_data );
 	}
 
+	/**
+	 * Sets up a handler to add error details to the payment result.
+	 * Registers an action to handle 'update_payment_result_on_error',
+	 * using the payment result object from 'woocommerce_rest_checkout_process_payment_with_context'.
+	 *
+	 * @param PaymentContext $context The payment context.
+	 * @param PaymentResult  $result  The payment result, passed by reference.
+	 */
+	public function setup_payment_error_handler( PaymentContext $context, PaymentResult &$result ) {
+		add_action(
+			'update_payment_result_on_error',
+			function ( $error ) use ( &$result ) {
+				$result->set_payment_details(
+					array_merge(
+						$result->payment_details,
+						[ 'errorMessage' => wp_strip_all_tags( $error->getMessage() ) ]
+					)
+				);
+			}
+		);
+	}
```

Comment:
I was considering pros and cons of this implementation, because of nested hooks. An alternative I considered was to keep the state of payment `$result` object in the local variable. However, this approach means we need to ensure the state of this variable is managed properly which could add a bit more complexity. So I ended up with this solution. More context: 

1. [Similarly implemented in Stripe Gateway](https://github.com/woocommerce/woocommerce-gateway-
stripe/blob/3376adee168dbe10689fef30e2346b93e8b0461c/includes/class-wc-stripe-blocks-support.php#L381-L388)
2. Consulted on this with Blocks team in p1719486293718779/1718284000.418279-slack-C02TS23QJ1X